packages/flutter/lib/src/material/menu_anchor.dart

```diff
@@ -596,6 +639,7 @@ _MenuAnchorState? get _previousFocusableSibling {
       // Notify that _childIsOpen changed state, but only if not
       // currently disposing.
       _parent?._childChangedOpenState();
+      await _animateController.reverse();
```

Comment:
Is it correct to `await` here instead of at the end of the method? The `widget.onClose` won't be called before the entire animation has ended.

Reply:
Also, do you think it would be a good idea to add a new callback called `onCloseAnimationEnd` or something? Imagine you're building a menu and you want to remove the `MenuAnchor` once an option is selected, but immediately removing it upon `onClose` (suppose we don't wait for the animation, and we shouldn't) might remove the fade out animation entirely (does it?)

Reply:
When working without access to the MenuAnchor internals, I found it easiest to just treat a closed menu as AnimationStatus.dismissed, and an open menu as anything but AnimationStatus.dismissed, and provide an `onAnimationStatusChange` callback. Right now, MenuController.isOpen just represents whether the overlay is showing, so it'd probably be least disruptive to keep it that way. 

With regard to `onAnimationStatusChange`, I prefer Tong's method of splitting out callbacks -- at one point, I considered doing `onAnimatingClosed`, `onAnimatingOpen`, `onClose`, `onOpen`, and will probably change it back. Note that the code below works with simulations, so regular animation code should be simpler.

Ideally, code that changes the internals of MenuAnchor will call _animateClosed when MenuController.close() is called, and _animateOpen when MenuController.open() is called.


```dart
  // Sets the menu status to closed and sets the menu animation to 0.0. Does not
  // trigger the root menu to close the overlay.
  void _handleClosed() {
    _animationController.stop();
    _animationController.value = 0;
    _updateMenuStatus(MenuStatus.closed);
  }

  // Sets the menu status to opened and sets the menu animation to 1.0. Does not
  // trigger the root menu to open the overlay.
  void _handleOpened() {
    _animationController.stop();
    _animationController.value = 1;
    _updateMenuStatus(MenuStatus.opened);
  }

  // Animate the menu closed, then trigger the root menu to close the overlay.
  void _animateClosed() {
    if (_menuStatus case MenuStatus.closed || MenuStatus.closing) {
      assert(_debugMenuInfo('Blocked $_animateClosed because the menu is already closing'));
      return;
    }

    // When the animation controller finishes closing, the inner menu's onClose
    // callback will be called, thereby triggering the _handleClosed callback.
    _animationController
      ..stop()
      ..animateWith(
        ClampedSimulation(
          SpringSimulation(
            widget.reverseSpring,
            _animationController.value,
            0.0,
            5.0,
            tolerance: _springTolerance,
          ),
          xMin: 0.0,
          xMax: 1.0,
        ),
      ).whenComplete(_innerMenuController.close); // I had to use two menu controllers to get around the fact that the inner controller would close the overlay when MenuController.close() was called. This is something that I've been trying to think through.

    _updateMenuStatus(MenuStatus.closing);
  }

  void _animateOpen({ui.Offset? position}) {
    if (_menuStatus case MenuStatus.opened || MenuStatus.opening) {
      _innerMenuController.open(position: position);
      _animationController.value = 1.0;
      return;
    }

    if (!_innerMenuController.isOpen) {
      _innerMenuController.open(position: position);
    }

    _animationController
      ..stop()
      ..animateWith(SpringSimulation(
        widget.forwardSpring,
        _animationController.value,
        1.0,
        5.0,
      )).whenComplete(_handleOpened);

    _updateMenuStatus(MenuStatus.opening);

    // When the menu is first opened, set the first focus to the first item in
    // the menu.
    SchedulerBinding.instance.addPostFrameCallback((Duration timeStamp) {
      final BuildContext? panelContext = _panelScrollableKey.currentContext;
      if (mounted && (panelContext?.mounted ?? false)) {
        FocusScope.of(context).setFirstFocus(
          FocusScope.of(panelContext!),
        );
      }
    });
  }

  // Update the menu status and call listeners.
  void _updateMenuStatus(MenuStatus status) {
    if (status == _menuStatus) {
      return;
    }

    final MenuStatus previousStatus = _menuStatus;
    _menuStatus = status;

    // Cannot use a postFrameCallback because focus won't return to the previous
    // focus node when the menu is closed.
    if (mounted && SchedulerBinding.instance.schedulerPhase !=
                   SchedulerPhase.persistentCallbacks) {
      setState(() { /* Mark dirty if mounted and not already building. */ });
    }

    if (previousStatus == MenuStatus.closed) {
      widget.onOpen?.call();
    }

    if (status == MenuStatus.closed) {
      widget.onClose?.call();
    }

    widget.onAnimationStatusChanged?.call(status);
  }
```